File: elixir/bob/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    cond do
      uppercase?(input) and input =~ ~r/.+\?/ -> "Calm down, I know what I'm doing!"
      uppercase?(input) -> "Whoa, chill out!"
      input =~ ~r/.+\?\s*$/ -> "Sure."
      input =~ ~r/^\s*$/ -> "Fine. Be that way!"
      true -> "Whatever."
    end
  end

  defp uppercase?(string) do
    string == String.upcase(string) and string != String.downcase(string) 
  end
end

File: ruby/bob/bob.rb
class Bob

  def self.hey(input)
    input = input.strip
    isYelling = (input.eql? input.upcase) && (not input.eql? input.downcase)
    isQuestion = input[-1].eql? "?"

    if isYelling and isQuestion
      "Calm down, I know what I'm doing!"
    elsif isYelling
      "Whoa, chill out!"
    elsif isQuestion
      "Sure."
    elsif input.empty?
      "Fine. Be that way!"
    else
      "Whatever."
    end
  end

end
